x-pack/plugins/cloud_security_posture/public/pages/rules/use_change_csp_rule_state.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { useKibana } from '@kbn/kibana-react-plugin/public';
+import { useQueryClient, useMutation } from '@tanstack/react-query';
+import { RULE_STATE_QUERY_KEY as RULES_STATE_QUERY_KEY } from './use_csp_rules_state';
+import { CSPM_STATS_QUERY_KEY, KSPM_STATS_QUERY_KEY } from '../../common/api';
+import { BENCHMARK_INTEGRATION_QUERY_KEY_V2 } from '../benchmarks/use_csp_benchmark_integrations';
+import {
+  CspBenchmarkRulesBulkActionRequestSchema,
+  RuleStateAttributes,
+} from '../../../common/types/latest';
+import { CSP_BENCHMARK_RULES_BULK_ACTION_ROUTE_PATH } from '../../../common/constants';
+
+export type RuleStateAttributesWithoutStates = Omit<RuleStateAttributes, 'muted'>;
+export interface RuleStateUpdateRequest {
+  newState: 'mute' | 'unmute';
+  ruleIds: RuleStateAttributesWithoutStates[];
+}
+
+export const useChangeCspRuleState = () => {
+  const { http } = useKibana().services;
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: async (ruleStateUpdateRequest: RuleStateUpdateRequest) => {
+      await http?.post<CspBenchmarkRulesBulkActionRequestSchema>(
+        CSP_BENCHMARK_RULES_BULK_ACTION_ROUTE_PATH,
+        {
+          version: '1',
+          body: JSON.stringify({
+            action: ruleStateUpdateRequest.newState,
+            rules: ruleStateUpdateRequest.ruleIds,
+          }),
+        }
+      );
+    },
+    onMutate: async (ruleStateUpdateRequest: RuleStateUpdateRequest) => {
+      // Cancel any outgoing refetches (so they don't overwrite our optimistic update)
+      await queryClient.cancelQueries(RULES_STATE_QUERY_KEY);
+
+      // Snapshot the previous value
+      const previousCspRules = queryClient.getQueryData(RULES_STATE_QUERY_KEY);
+
+      // Optimistically update to the new value
+      queryClient.setQueryData(['csp_rules_states_v1'], (currentRuleStates: any) => {
+        return createRulesWithUpdatedState(ruleStateUpdateRequest, currentRuleStates);
+      });
+
+      // Return a context object with the previous value
+      return { previousCspRules };
```

Comment:
we should also use this data which is being passed as `context` on an `onError` property to revert the cache to its previous state in case something goes wrong

Reply:
Should be there now.